src/esp/gfx/Renderer.cpp

```diff
@@ -74,6 +70,51 @@ struct Renderer::Impl {
   void draw(RenderCamera& camera, MagnumDrawableGroup& drawables) {
     renderEnter();
     camera.getMagnumCamera().setViewport(framebufferSize_);
+
+    /* Inverted projection matrix to unproject the depth value and chop the
+       near plane off. We don't care about X/Y there and the corresponding
+       parts of the matrix are zero as well so take just the lower-right part
+       of it (denoted a, b, c, d).
+
+        x 0 0 0
+        0 y 0 0
+        0 0 a b
+        0 0 c d
+
+      Taking a 2-component vector with the first component being Z and second
+      1, the final calculation of unprojected Z is then
+
+        | a b |   | z |   | az + b |
+        | c d | * | 1 | = | cz + d |
+
+    */
+    const Matrix4 projection = camera.getMagnumCamera().projectionMatrix();
+    const Matrix4 invertedProjection =
+        Matrix4::translation(Vector3::zAxis(-projection.translation().z())) *
```

Comment:
I think this is where you are chopping off the near-plane by moving the near-plane to z=0. How does it work? Why would projection.translation().z() give you the correct z-translation?

Reply:
Hmm. :thinking: Good point, actually :+1: 

I won't lie -- without this, the output was different from the previous depth value written explicitly by the shader by about 0.02. And the projection translation Z is also about 0.02, and this operation seemed to make sense at the time -- that was even before the tests were passing, as I hadn't yet discovered the discrepancy between mesh holes being represented as 0 in the previous case and 1 by the GPU depth buffer, and thus wasn't aware that this shift doesn't have much effect on the test passing. It passes both with and without, the threshold is large enough to allow that.

Now, when going through the actual matrix math, it seems totally arbitrary and just a lucky coincidence :sweat_smile: Furthermore, when removing this arbitrary shift, the resulting depth frame is much closer to the actual ground truth (could reduce the threshold to 0.3e-2 from 1.5e-2 and it still passes). Removing this, then, thanks a lot for the heads-up.

A remaining mystery is the 0.02 difference between values produced by the GPU depth buffer and by the explicit shader calculation -- but since I'm closer to the ground truth with the GPU depth buffer, I would assume it's due to some rounding errors happening at various places (and also the switch from 24-bit depth to 32F depth could have an effect on that).